File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();

        int s1 = 0, s2 = 0, e1 = matrix.length-1, e2 = matrix[0].length-1;
        while(s1<=e1 && s2<=e2){
            for(int i = s2; i<=e2; i++){
                res.add(matrix[s1][i]);
            }

            for(int i=s1+1; i<=e1; i++){
                res.add(matrix[i][e2]);
            }

            if(s1<e1){
                for(int i=e2-1; i>=s2; i--){
                    res.add(matrix[e1][i]);
                }
            }

            if(s2<e2){
                for(int i=(e1-1); i>s1; i--){
                    res.add(matrix[i][s1]);
                }
            }

            s1++; s2++; e2--; e1--;
        }

        return res;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []

        s1, s2, e1, e2 = 0, 0, len(matrix)-1, len(matrix[0])-1

        while(s1<=e1 and s2<=e2):
            for i in range(s2, e2+1):
                res.append(matrix[s1][i])
            
            for i in range(s1+1, e1+1):
                res.append(matrix[i][e2])
            
            if(s1<e1):
                for i in range(e2-1, s2-1, -1):
                    res.append(matrix[e1][i])
            
            if(s2<e2):
                for i in range(e1-1, s1, -1):
                    res.append(matrix[i][s1])
            
            s1+=1
            s2+=1
            e1-=1
            e2-=1
        
        return res